### PR TITLE
Fixed: Kubescape fails to authenticate remote private Github repo

### PR DESCRIPTION
**Pull request by suhasgumma**
## Describe your changes
* Checked if the Git Repository is Public or Private.
* No Authentication is required for scanning Public Git Repositories
* Added authentication using GITHUB_TOKEN for scanning private Git repositories

## Issue ticket number and link
 #710

## Checklist before requesting a review
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have performed a self-review of my code
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: core/pkg/resourcehandler/remotegitutils.go [hunk 1]
```diff
@@ -18,9 +38,26 @@ func cloneRepo(gitURL giturl.IGitURL) (string, error) {
...
+	// Get the URL to clone
 	cloneURL := gitURL.GetHttpCloneURL()
-	cloneOpts := git.CloneOptions{URL: cloneURL}
+
+	isGitRepoPublic := isGitRepoPublic(cloneURL)
+
+	// Declare the authentication variable required for cloneOptions
+	var auth transport.AuthMethod
+
+	if isGitRepoPublic {
```
Comment (dwertent) [inline core/pkg/resourcehandler/remotegitutils.go:49]: Is it necessary to pre-check if the repo is public or private? What if you pass an empty token?

File: core/pkg/resourcehandler/remotegitutils.go [hunk 2]
```diff
@@ -18,9 +38,26 @@ func cloneRepo(gitURL giturl.IGitURL) (string, error) {
...
+		// No authentication needed if repository is public
+		auth = nil
+	} else {
+		auth = &http.BasicAuth{
```
Comment (dwertent) [inline core/pkg/resourcehandler/remotegitutils.go:53]: We should check if there is a token- if not, the function should return an error

File: core/pkg/resourcehandler/remotegitutils.go [hunk 3]
```diff
@@ -18,9 +38,26 @@ func cloneRepo(gitURL giturl.IGitURL) (string, error) {
...
+			Username: "anything Except Empty String",
+			Password: os.Getenv("GITHUB_TOKEN"),
```
Comment (dwertent) [inline core/pkg/resourcehandler/remotegitutils.go:55]: We should use the `giturl.IGitURL` interface for loading the token.
We might need to add a `GetToken` method to the interface

**Review by dwertent — CHANGES_REQUESTED** [review]

File: core/pkg/resourcehandler/remotegitutils.go (hunk 3 above)
Comment (suhasgumma) [reply core/pkg/resourcehandler/remotegitutils.go:55]: * There is another interface called `giturl.IGitAPI` which inherits `giturl.IGitURL` has some additional methods including `GetToken`. 
* I tested the code using that interface (only small changes required) and the code is working fine. Shall we proceed with that?

Comment (suhasgumma) [reply core/pkg/resourcehandler/remotegitutils.go:55]: * There is another interface called `giturl.IGitAPI` which inherits `giturl.IGitURL` has some additional methods including `GetToken`.
* I tested the code using that interface (only small changes required) and the code is working fine. Shall we proceed with that?

File: core/pkg/resourcehandler/remotegitutils.go (hunk 1 above)
Comment (suhasgumma) [reply core/pkg/resourcehandler/remotegitutils.go:49]: * If we are using `auth` in cloneOptions, it is throwing an error even for public repositories if the token is empty.
* In my understanding, we have to implement different authentication logic for public and private repositories. So, we should definitely pre-check if the repo is public or private. Correct me if I am wrong.

File: core/pkg/resourcehandler/remotegitutils.go (hunk 3 above)
Comment (pwnb0y) [reply core/pkg/resourcehandler/remotegitutils.go]: I think it is not a secure way to authenticate a user by setting their token in the environment.
I think one switch is also available like -p by using that flag users can pass their token in script or it prompts input for the token,
if someone put their token in their os environment they can be exploited.

**Review by dwertent — APPROVED** [review]

**Review by dwertent — APPROVED** [review]

**Pull request merged**